src/borg/crypto/key.py

```diff
@@ -432,13 +436,41 @@ def decrypt_key_file(self, data, passphrase):
         unpacker = get_limited_unpacker('key')
         unpacker.feed(data)
         data = unpacker.unpack()
-        enc_key = EncryptedKey(internal_dict=data)
-        assert enc_key.version == 1
-        assert enc_key.algorithm == 'sha256'
-        key = passphrase.kdf(enc_key.salt, enc_key.iterations, 32)
-        data = AES(key, b'\0'*16).decrypt(enc_key.data)
-        if hmac_sha256(key, data) == enc_key.hash:
+        encrypted_key = EncryptedKey(internal_dict=data)
+        if encrypted_key.version == 1:
+            return self.decrypt_key_file_v1(encrypted_key, passphrase)
+        elif encrypted_key.version == 2:
+            return self.decrypt_key_file_v2(encrypted_key, passphrase)
+        else:
+            raise UnsupportedKeyFormatError()
+
+    def decrypt_key_file_v1(self, encrypted_key, passphrase):
+        if encrypted_key.algorithm != 'sha256':
+            raise UnsupportedKeyFormatError()
+        key = passphrase.kdf(encrypted_key.salt, encrypted_key.iterations, 32)
+        data = AES(key, b'\0'*16).decrypt(encrypted_key.data)
+        if hmac_sha256(key, data) == encrypted_key.hash:
             return data
+        return None
+
+    def decrypt_key_file_v2(self, encrypted_key, passphrase):
+        if encrypted_key.algorithm != "argon2 aes256-ctr hmac-sha256":
+            raise UnsupportedKeyFormatError()
+        key = passphrase.argon2(
+            output_len_in_bytes=64,
+            salt=encrypted_key.salt,
+            time_cost=encrypted_key.argon2_time_cost,
+            memory_cost=encrypted_key.argon2_memory_cost,
+            parallelism=encrypted_key.argon2_parallelism,
+            type=encrypted_key.argon2_type,
+        )
+        enc_key, mac_key = key[:32], key[32:]
+        algorithm = AES256_CTR_HMAC_SHA256(
+            iv=0, header_len=0, aad_offset=0,
+            enc_key=enc_key,
+            mac_key=mac_key,
+        )
+        return algorithm.decrypt(encrypted_key.data)
```

Comment:
473 has different behaviour (throws exception on failure) than 452..454 (returns None on failure).

Reply:
did you see this ^^^?

Reply:
I did. This is causing [the passphrase retry bug I mentioned](https://github.com/borgbackup/borg/pull/6469#issuecomment-1079898226). I'll fix this.